**Loom video link: https://www.loom.com/share/f1eda7ca5c134abdbf668ff611b87bb4**

**Github Repo: https://github.com/philip41201/Tracking-Changes-in-Demand-and-Pricing-for-Live-Events-**

In [ ]:
!jupyter nbconvert presentation.ipynb --to slides --post serve

[NbConvertApp] Converting notebook presentation.ipynb to slides
[NbConvertApp] Writing 647980 bytes to presentation.slides.html
[NbConvertApp] Redirecting reveal.js requests to https://cdnjs.cloudflare.com/ajax/libs/reveal.js/3.5.0
Serving your slides at http://127.0.0.1:8000/presentation.slides.html
Use Control-C to stop this server


# Ticketmaster Data Collection 

In [ ]:
import requests
import json
import pandas as pd
from sqlalchemy import create_engine

In [2]:
api_url = 'https://app.ticketmaster.com/discovery/v2/events.json'
params = {'apikey':'API_KEY', 'keyword':'Garth Brooks', 'countryCode':'US'}
api_request = requests.get(api_url, params=params)

In [3]:
api_request

<Response [200]>

In [4]:
api_request.url

'https://app.ticketmaster.com/discovery/v2/events.json?apikey=g41Q8ILou71Xrbkop2xuNXrS5pBo53hv&keyword=Garth+Brooks&countryCode=US'

In [5]:
details = json.loads(api_request.text)['_embedded']['events']

In [6]:
event_details = {
    'event_id':[],
    'name':[],
    'genre':[],
    'venue':[],
    'city':[],
    'state':[],
    'publicOnSaleDate':[],
    'eventDate':[],
    'minPrice':[],
    'maxPrice':[],
}

for event in details:
    event_id = event['id']
    event_details['event_id'].append(event_id)
    print(event_id)
    
    name = event['name']
    event_details['name'].append(name)
    print(name)
    
    genre = event['classifications'][0]['genre']['name']
    event_details['genre'].append(genre)
    print(genre)
    
    venue = event['_embedded']['venues'][0]['name']
    event_details['venue'].append(venue)
    print(venue)
    
    city = event['_embedded']['venues'][0]['city']['name']
    event_details['city'].append(city)
    print(city)
    
    state = event['_embedded']['venues'][0]['state']['name']
    event_details['state'].append(state)
    print(state)
    
    publicOnSaleDate = event['sales']['public']['startDateTime'][0:10]
    event_details['publicOnSaleDate'].append(publicOnSaleDate)
    print(publicOnSaleDate)
    
    eventDate = event['dates']['start']['localDate']
    event_details['eventDate'].append(eventDate)
    print(eventDate)

    minPrice = event['priceRanges'][0]['min']
    event_details['minPrice'].append(minPrice)
    print(minPrice)
    
    maxPrice = event['priceRanges'][0]['max']
    event_details['maxPrice'].append(maxPrice)
    print(maxPrice)
 
'''
event_details['event_id'].append(details[6]['id'])
event_details['name'].append(details[6]['name'])
event_details['genre'].append(details[6]['classifications'][0]['genre']['name'])
event_details['venue'].append(details[6]['_embedded']['venues'][0]['name'])
event_details['city'].append(details[6]['_embedded']['venues'][0]['city']['name'])
event_details['state'].append(details[6]['_embedded']['venues'][0]['state']['name'])
event_details['publicOnSaleDate'].append(details[6]['sales']['public']['startDateTime'][0:10])
event_details['eventDate'].append(details[6]['dates']['start']['localDate'])
event_details['minPrice'].append(details[6]['priceRanges'][0]['min'])
event_details['maxPrice'].append(details[6]['priceRanges'][0]['max'])
'''

vvG1YZ9kLFAxqp
Garth Brooks
Country
AT&T Stadium
Arlington
Texas
2022-04-15
2022-07-30
91.1
91.1
vvG1fZ97UTZwF6
Garth Brooks
Country
Notre Dame Stadium
South Bend
Indiana
2022-02-25
2022-05-07
91.7
91.7
G5vbZ9AS6qkCd
Garth Brooks
Country
Paul Brown Stadium
Cincinnati
Ohio
2022-03-11
2022-05-14
84.68
84.68
G5eZZ96tZ15UN
Garth Brooks
Country
Protective Stadium
Birmingham
Alabama
2022-04-08
2022-06-04
91.7
91.7
vvG17Z9ewYmxCz
Garth Brooks
Country
Gillette Stadium
Foxborough
Massachusetts
2021-02-14
2022-05-21
84.74
84.74
G5vzZ9Fh0m0RW
Garth Brooks
Country
Rice-Eccles Stadium
Salt Lake City
Utah
2022-04-22
2022-06-18
87.7
87.7
G5vzZ9aG3m_dN
Garth Brooks
Country
Rice-Eccles Stadium
Salt Lake City
Utah
2022-04-29
2022-06-17
87.7
87.7
G5vbZ9FRVJfP5
Garth Brooks
Country
Paul Brown Stadium
Cincinnati
Ohio
2022-04-29
2022-05-13
84.68
84.68
G5eVZ965LN9d7
Garth Brooks
Country
Bank of America Stadium
Charlotte
North Carolina
2022-03-25
2022-07-15
79.41
79.41
G5eVZ9Ar8F3ZY
Garth Brooks
Country
Bank 

"\nevent_details['event_id'].append(details[6]['id'])\nevent_details['name'].append(details[6]['name'])\nevent_details['genre'].append(details[6]['classifications'][0]['genre']['name'])\nevent_details['venue'].append(details[6]['_embedded']['venues'][0]['name'])\nevent_details['city'].append(details[6]['_embedded']['venues'][0]['city']['name'])\nevent_details['state'].append(details[6]['_embedded']['venues'][0]['state']['name'])\nevent_details['publicOnSaleDate'].append(details[6]['sales']['public']['startDateTime'][0:10])\nevent_details['eventDate'].append(details[6]['dates']['start']['localDate'])\nevent_details['minPrice'].append(details[6]['priceRanges'][0]['min'])\nevent_details['maxPrice'].append(details[6]['priceRanges'][0]['max'])\n"

In [7]:
event_df = pd.DataFrame(event_details)

In [8]:
event_df.head()

event_id          name    genre               venue        city  \
0  vvG1YZ9kLFAxqp  Garth Brooks  Country        AT&T Stadium   Arlington   
1  vvG1fZ97UTZwF6  Garth Brooks  Country  Notre Dame Stadium  South Bend   
2   G5vbZ9AS6qkCd  Garth Brooks  Country  Paul Brown Stadium  Cincinnati   
3   G5eZZ96tZ15UN  Garth Brooks  Country  Protective Stadium  Birmingham   
4  vvG17Z9ewYmxCz  Garth Brooks  Country    Gillette Stadium  Foxborough   

           state publicOnSaleDate   eventDate  minPrice  maxPrice  
0          Texas       2022-04-15  2022-07-30     91.10     91.10  
1        Indiana       2022-02-25  2022-05-07     91.70     91.70  
2           Ohio       2022-03-11  2022-05-14     84.68     84.68  
3        Alabama       2022-04-08  2022-06-04     91.70     91.70  
4  Massachusetts       2021-02-14  2022-05-21     84.74     84.74

In [ ]:
engine = create_engine('mysql+mysqldb://USERNAME:PASSWORD@HOST/DATABASE?charset=utf8')

In [425]:
event_df.to_sql('ticketmaster_event_details', engine, if_exists='append', index=False)

# Venue Data Collection

In [1]:
import requests
import pandas as pd
import json
from datetime import datetime, timezone
from sqlalchemy import create_engine
from bs4 import BeautifulSoup

In [2]:
venue = input("What venue are you searching for? ")

What venue are you search?ing for? Hollywood Bowl


In [3]:
url = 'https://www.google.com/search'
params = {'q':venue + 'capacity'}
headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.75 Safari/537.36'}

In [4]:
event_request = requests.get(url, params=params, headers=headers)

In [5]:
soup = BeautifulSoup(event_request.text, 'html.parser')

In [6]:
capacity = soup.find('div', attrs={'class':'Z0LcW'})

In [7]:
capacity.text

'17,500'

In [8]:
url = 'https://www.google.com/search'
params = {'q':venue}
headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.75 Safari/537.36'}

In [9]:
event_request = requests.get(url, params=params, headers=headers)

In [10]:
soup = BeautifulSoup(event_request.text, 'html.parser')

In [11]:
address = soup.find('span', attrs={'class':'LrzXr'})

In [12]:
address.text

'2301 N Highland Ave, Los Angeles, CA 90068'

In [14]:
address.text.split(',')

['2301 N Highland Ave', ' Los Angeles', ' CA 90068']

In [17]:
address.text.split(',')[2][1:3]

'CA'

In [18]:
venue_details = {
    'name':[],
    'address':[],
    'city':[],
    'state':[],
    'capacity':[]
}

venue_details['name'].append(venue)
venue_details['address'].append(address.text.split(',')[0])
venue_details['city'].append(address.text.split(',')[1])
venue_details['state'].append(address.text.split(',')[2][1:3])
venue_details['capacity'].append(capacity.text)

In [19]:
venue_details

{'name': ['Hollywood Bowl'],
 'address': ['2301 N Highland Ave'],
 'city': [' Los Angeles'],
 'state': ['CA'],
 'capacity': ['17,500']}

In [833]:
venue_df = pd.DataFrame(venue_details)

In [ ]:
engine = create_engine('mysql+mysqldb://USERNAME:PASSWORD@HOST/DATABASE?charset=utf8')

In [834]:
venue_df.to_sql('venues', engine, if_exists='append', index=False)

# Vividseats Data Collection

In [56]:
import requests
import pandas as pd
import json
from datetime import datetime, timezone
from sqlalchemy import create_engine
from bs4 import BeautifulSoup

In [57]:
event_id = ['3855914','3897660','3912847','3838061','3873578','3869589','3855658','3845915','3855657','3849613','3846005']
headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.75 Safari/537.36'}

In [58]:
event_url = 'https://www.vividseats.com/production/'
urls = []
for event in event_id:
    url = event_url+event
    urls.append(url)

In [59]:
event_requests = []

for url in urls:
    event_request = requests.get(url, headers=headers)
    er = event_request.text
    event_requests.append(er)

In [60]:
soups = []

for er in event_requests:
    soup = BeautifulSoup(er, 'html.parser')
    event_info = soup.find('script', attrs={'type':'application/json'})
    ei = json.loads(event_info.text)
    soups.append(ei)


In [61]:
detail = []

for ei in soups:
    details = ei['props']['pageProps']['initialProductionDetailsData']['data']
    detail.append(details)

In [62]:
event_details = {
    'name':[],
    'venue':[],
    'vs_event_id':[],
    'listingCount':[],
    'ticketCount':[],
    'minPrice':[],
    'maxPrice':[],
    'eventDate':[],
    'currentDate':[]
}

for event in detail:
    event_details['name'].append(event['name'])
    event_details['venue'].append(event['venue']['name'])
    event_details['vs_event_id'].append(event['id'])
    event_details['listingCount'].append(event['listingCount'])
    event_details['ticketCount'].append(event['ticketCount'])
    event_details['minPrice'].append(event['minPrice'])
    event_details['maxPrice'].append(event['maxPrice'])
    event_details['eventDate'].append(event['localDate'][0:10])
    event_details['currentDate'].append(datetime.now(timezone.utc).strftime("%Y-%m-%d"))

In [63]:
event_df = pd.DataFrame(event_details)

In [64]:
event_df

name  \
0   Keshi (Moved from Showbox at the Market)   
1                                        UMI   
2                                 Chris Rock   
3                              Jimmy Buffett   
4                          HER - Gabi Wilson   
5                             Paul McCartney   
6                                      Keshi   
7                          Rex Orange County   
8                                      Keshi   
9                         Dave Matthews Band   
10                         Rex Orange County   

                                           venue  vs_event_id  listingCount  \
0                                   Showbox SODO      3855914            45   
1                      The Observatory Santa Ana      3897660            63   
2                                   Taft Theatre      3912847           261   
3                               Moody Center ATX      3838061           683   
4                      Dailys Place Amphitheater      3873578           228   
5                                MetLife Stadium      3869589          1023   
6                      The Observatory Santa Ana      3855658            52   
7                            BMO Harris Pavilion      3845915           299   
8                     The Observatory North Park      3855657            42   
9               Ruoff Home Mortgage Music Center      3849613           227   
10  Huntington Bank Pavilion at Northerly Island      3846005           346   

    ticketCount  minPrice  maxPrice   eventDate currentDate  
0           212       103      7502  2022-06-09  2022-04-21  
1           229        41       224  2022-06-09  2022-04-21  
2           686        60      1604  2022-06-10  2022-04-21  
3          2507        16     19626  2022-06-11  2022-04-21  
4          1516        40      1053  2022-06-12  2022-04-21  
5          3180       107      8422  2022-06-16  2022-04-21  
6           248        95       674  2022-06-17  2022-04-21  
7          1403        44       758  2022-06-17  2022-04-21  
8           207        94       674  2022-06-18  2022-04-21  
9          1119        63      1238  2022-06-24  2022-04-21  
10         1147        50      8421  2022-06-24  2022-04-21

In [65]:
engine = create_engine('mysql+mysqldb://USERNAME:PASSWORD@HOST/DATABASE?charset=utf8')

In [66]:
event_df.to_sql('event_details', engine, if_exists='append', index=False)

# SQL Analysis

In [2]:
%load_ext sql

In [3]:
%sql mysql://admin:sql_2022@tickets.ceaw7c57s0vu.us-east-1.rds.amazonaws.com/tickets

### 1. This query sums how many events I have in my Ticketmaster table by genre.

In [6]:
%%sql
SELECT
    COUNT(event_id) AS number_of_events,
    genre
FROM ticketmaster_event_details
GROUP BY genre
ORDER BY number_of_events DESC;

 * mysql://admin:***@tickets.ceaw7c57s0vu.us-east-1.rds.amazonaws.com/tickets
7 rows affected.


[(29, 'Rock'),
 (16, 'Country'),
 (6, 'Pop'),
 (5, 'Comedy'),
 (5, 'Other'),
 (2, 'R&B'),
 (1, 'Hip-Hop/Rap')]

### I have majority rock and country events stored in my Ticketmaster table

### 2. This query compares ticket count for an event to the capacity of the venue by joining the vividseats and venue tables. 

In [8]:
%%sql
SELECT
    vi.name,
    venue,
    ticketCount,
    capacity
FROM vividseats_event_details vi
JOIN venue ve
    ON vi.venue = ve.name
LIMIT 10;

 * mysql://admin:***@tickets.ceaw7c57s0vu.us-east-1.rds.amazonaws.com/tickets
10 rows affected.


[('BTS - Bangtan Boys', 'Allegiant Stadium', 2135, '65,000'),
 ('Bon Jovi', 'Enmarket Arena', 1112, '9,500'),
 ('Jawbreaker', 'House of Blues Chicago', 145, '1,800'),
 ('Clinton Kane', 'Marquis Theater - CO', 30, '250'),
 ('Dave Chappelle', 'Uptown Theater Kansas City', 209, '1,700'),
 ('BTS - Bangtan Boys', 'Allegiant Stadium', 2030, '65,000'),
 ('Garth Brooks', 'Nissan Stadium', 2615, '69,143'),
 ('Yeat', 'Irving Plaza', 103, '1,200'),
 ('Dave Chappelle', 'Vivint Smart Home Arena', 2291, '18,300'),
 ('Dave Chappelle', 'Vivint Smart Home Arena', 2291, '18,300')]

### One of my questions could look into the ratio between ticketCount and capacity

### 3. This query looks at the ticket count and minimum price as number of days until the event changes.

In [12]:
%%sql
SELECT
    name,
    venue,
    ticketCount,
    minPrice,
    eventDate,
    eventDate - currentDate AS days_until_event
FROM vividseats_event_details
ORDER BY eventDate, name, days_until_event DESC
LIMIT 20;

 * mysql://admin:***@tickets.ceaw7c57s0vu.us-east-1.rds.amazonaws.com/tickets
20 rows affected.


[('BTS - Bangtan Boys', 'Allegiant Stadium', 2135, 54, datetime.date(2022, 4, 8), 0),
 ('BTS - Bangtan Boys', 'Allegiant Stadium', 1929, 60, datetime.date(2022, 4, 8), 0),
 ('Bon Jovi', 'Enmarket Arena', 1112, 49, datetime.date(2022, 4, 13), 5),
 ('Bon Jovi', 'Enmarket Arena', 924, 62, datetime.date(2022, 4, 13), 2),
 ('Bon Jovi', 'Enmarket Arena', 885, 49, datetime.date(2022, 4, 13), 1),
 ('Bon Jovi', 'Enmarket Arena', 873, 48, datetime.date(2022, 4, 13), 1),
 ('Jawbreaker', 'House of Blues Chicago', 145, 22, datetime.date(2022, 4, 14), 6),
 ('Jawbreaker', 'House of Blues Chicago', 105, 6, datetime.date(2022, 4, 14), 3),
 ('Jawbreaker', 'House of Blues Chicago', 87, 6, datetime.date(2022, 4, 14), 2),
 ('Jawbreaker', 'House of Blues Chicago', 75, 6, datetime.date(2022, 4, 14), 2),
 ('Jawbreaker', 'House of Blues Chicago', 38, 17, datetime.date(2022, 4, 14), 0),
 ('BTS - Bangtan Boys', 'Allegiant Stadium', 2030, 80, datetime.date(2022, 4, 15), 7),
 ('BTS - Bangtan Boys', 'Allegiant Stadium', 1346, 113, datetime.date(2022, 4, 15), 4),
 ('BTS - Bangtan Boys', 'Allegiant Stadium', 1133, 134, datetime.date(2022, 4, 15), 3),
 ('BTS - Bangtan Boys', 'Allegiant Stadium', 1145, 134, datetime.date(2022, 4, 15), 3),
 ('BTS - Bangtan Boys', 'Allegiant Stadium', 804, 149, datetime.date(2022, 4, 15), 1),
 ('Clinton Kane', 'Marquis Theater - CO', 30, 34, datetime.date(2022, 4, 15), 7),
 ('Clinton Kane', 'Marquis Theater - CO', 38, 29, datetime.date(2022, 4, 15), 4),
 ('Clinton Kane', 'Marquis Theater - CO', 22, 29, datetime.date(2022, 4, 15), 3),
 ('Clinton Kane', 'Marquis Theater - CO', 26, 29, datetime.date(2022, 4, 15), 3)]

### I could look into changes in ticket count and minimum price to gauge demand as the event comes closer to determine whether it's better to sell earlier or later.

### 4. Compare all face value minimum price and resale minimum price to get an idea of whether an event is profitable.

In [17]:
%%sql
SELECT
    ted.name,
    ted.minPrice AS retail_min_price,
    ved.minPrice AS resale_min_price
FROM ticketmaster_event_details ted
JOIN venue
    ON venue.name = ted.venue
JOIN vividseats_event_details ved
    ON venue.name = ved.venue;

 * mysql://admin:***@tickets.ceaw7c57s0vu.us-east-1.rds.amazonaws.com/tickets
307 rows affected.


[('Kane Brown - Blessed & Free Tour', 56, 49),
 ('mxmtoon - rising (the tour) ', 28, 22),
 ('keshi: HELL / HEAVEN TOUR', 35, 22),
 ('Kane Brown - Blessed & Free Tour', 56, 62),
 ('mxmtoon - rising (the tour) ', 28, 6),
 ('keshi: HELL / HEAVEN TOUR', 35, 6),
 ('Dave Matthews Band', 60, 103),
 ('iHeartCountry Festival Presented by Capital One', 29, 103),
 ('Jimmy Buffett', 54, 103),
 ('Netflix Is A Joke Presents: Bill Burr & Friends', 64, 75),
 ('Netflix Is A Joke Presents: Bill Burr & Friends', 64, 75),
 ("Clinton Kane - MAYBE SOMEDAY IT'LL ALL BE OK", 25, 32),
 ('Netflix Is A Joke Presents: Bill Burr: Slight Return', 60, 138),
 ('Pearl Jam', 108, 138),
 ('Pearl Jam', 108, 138),
 ('PAUL McCARTNEY GOT BACK', 30, 127),
 ('Dave Matthews Band', 60, 99),
 ('iHeartCountry Festival Presented by Capital One', 29, 99),
 ('Jimmy Buffett', 54, 99),
 ('Garth Brooks', 92, 84),
 ('mxmtoon - rising (the tour) ', 28, 15),
 ('keshi: HELL / HEAVEN TOUR', 35, 15),
 ('Netflix Is A Joke Presents: Bill Burr: Slight Return', 60, 148),
 ('Pearl Jam', 108, 148),
 ('Pearl Jam', 108, 148),
 ('Dave Matthews Band', 60, 71),
 ('iHeartCountry Festival Presented by Capital One', 29, 71),
 ('Jimmy Buffett', 54, 71),
 ('Garth Brooks', 85, 89),
 ('Garth Brooks', 85, 89),
 ('Dave Matthews Band', 50, 77),
 ('TWICE 4TH WORLD TOUR  III  IN NORTH AMERICA ENCORE', 60, 100),
 ('TWICE 4TH WORLD TOUR  III  IN NORTH AMERICA ENCORE', 60, 100),
 ('Dave Matthews Band', 50, 108),
 ('PAUL McCARTNEY GOT BACK', 40, 404),
 ('Rex Orange County: The Who Cares? Tour', 30, 57),
 ('Luke Combs Stadium Tour', 21, 76),
 ('Garth Brooks', 85, 49),
 ('Garth Brooks', 85, 49),
 ('Kane Brown - Blessed & Free Tour', 47, 57),
 ('Kane Brown - Blessed & Free Tour', 56, 52),
 ('Kane Brown - Blessed & Free Tour', 56, 49),
 ('mxmtoon - rising (the tour) ', 28, 6),
 ('keshi: HELL / HEAVEN TOUR', 35, 6),
 ('Kane Brown - Blessed & Free Tour', 56, 48),
 ('mxmtoon - rising (the tour) ', 28, 6),
 ('keshi: HELL / HEAVEN TOUR', 35, 6),
 ('Dave Matthews Band', 60, 97),
 ('iHeartCountry Festival Presented by Capital One', 29, 97),
 ('Jimmy Buffett', 54, 97),
 ('mxmtoon - rising (the tour) ', 28, 17),
 ('keshi: HELL / HEAVEN TOUR', 35, 17),
 ('Dave Matthews Band', 60, 103),
 ('iHeartCountry Festival Presented by Capital One', 29, 103),
 ('Jimmy Buffett', 54, 103),
 ('Dave Matthews Band', 60, 97),
 ('iHeartCountry Festival Presented by Capital One', 29, 97),
 ('Jimmy Buffett', 54, 97),
 ('Netflix Is A Joke Presents: Bill Burr & Friends', 64, 63),
 ('Netflix Is A Joke Presents: Bill Burr & Friends', 64, 63),
 ('Netflix Is A Joke Presents: Bill Burr & Friends', 64, 63),
 ('Netflix Is A Joke Presents: Bill Burr & Friends', 64, 63),
 ("Clinton Kane - MAYBE SOMEDAY IT'LL ALL BE OK", 25, 55),
 ('PAUL McCARTNEY GOT BACK', 30, 125),
 ('Netflix Is A Joke Presents: Bill Burr: Slight Return', 60, 119),
 ('Pearl Jam', 108, 119),
 ('Pearl Jam', 108, 119),
 ('Garth Brooks', 92, 77),
 ('Dave Matthews Band', 60, 99),
 ('iHeartCountry Festival Presented by Capital One', 29, 99),
 ('Jimmy Buffett', 54, 99),
 ('mxmtoon - rising (the tour) ', 28, 8),
 ('keshi: HELL / HEAVEN TOUR', 35, 8),
 ('Netflix Is A Joke Presents: Bill Burr: Slight Return', 60, 127),
 ('Pearl Jam', 108, 127),
 ('Pearl Jam', 108, 127),
 ('Dave Matthews Band', 60, 72),
 ('iHeartCountry Festival Presented by Capital One', 29, 72),
 ('Jimmy Buffett', 54, 72),
 ('Dave Matthews Band', 50, 73),
 ('Garth Brooks', 85, 83),
 ('Garth Brooks', 85, 83),
 ('TWICE 4TH WORLD TOUR  III  IN NORTH AMERICA ENCORE', 60, 120),
 ('TWICE 4TH WORLD TOUR  III  IN NORTH AMERICA ENCORE', 60, 120),
 ('Dave Matthews Band', 50, 108),
 ('PAUL McCARTNEY GOT BACK', 40, 404),
 ('Rex Orange County: The Who Cares? Tour', 30, 56),
 ('Garth Brooks', 85, 49),
 ('Garth Brooks', 85, 49),
 ('Kane Brown - Blessed & Free Tour', 47, 62),
 ('Luke Combs Stadium Tour', 21, 74),
 ('Kane Brown - Blessed & Free Tour', 56, 48),
 ('Nine Inch Nails', 50, 90),
 ('Nine Inch Nails', 50, 90),
 ('keshi', 30, 108),
 ('C

### Able to join all 3 tables successfully and pull the data I need.

### 5. Sum the amount of venues we have stored grouped by state.

In [19]:
%%sql
SELECT
    COUNT(ID) AS number_of_venues,
    state
FROM venue
GROUP BY state
ORDER BY number_of_venues DESC
LIMIT 10;

 * mysql://admin:***@tickets.ceaw7c57s0vu.us-east-1.rds.amazonaws.com/tickets
10 rows affected.


[(9, 'California'),
 (8, 'Texas'),
 (4, 'Florida'),
 (3, 'Illinois'),
 (3, 'Colorado'),
 (3, 'Tennessee'),
 (3, 'Pennsylvania'),
 (3, 'Massachusetts'),
 (2, 'Ohio'),
 (2, 'Indiana')]

### We could possibly perform some analytics based on state or region.

# Primary Question
## When is the best time to sell a ticket?
### Deciding when to sell your ticket is one of the most important questions to answer. Being able to view changes in minimum price (get-in price) and ticket count as the event comes closer can be used to make pricing decisions that will help maximize profit.
#### SQL features used: VIEW, CTE, Window function

In [10]:
%%sql
CREATE VIEW charli_xcx_events AS
    SELECT *
    FROM vividseats
    WHERE name = 'Charli XCX'
ORDER BY eventDate;

 * mysql://admin:***@tickets.ceaw7c57s0vu.us-east-1.rds.amazonaws.com/tickets
0 rows affected.


[]

In [11]:
%%sql
WITH charli_xcx_boston AS (
    SELECT
        name,
        venue,
        eventDate - reportDate AS Days_Until_Event,
        ticketCount,
        LAG(ticketCount, 1) OVER(
            ORDER BY reportDate
        ) AS previous_ticket_count,
        (
            (ticketCount - LAG(ticketCount, 1) OVER(ORDER BY reportDate))
            /
            LAG(ticketCount, 1) OVER(ORDER BY reportDate)
        ) * 100 AS percent_change_ticket_count,
        minPrice,
        LAG(minPrice, 1) OVER(
            ORDER BY reportDate
        ) AS previous_min_price,
        (
            (minPrice - LAG(minPrice, 1) OVER(ORDER BY reportDate))
            /
            LAG(minPrice, 1) OVER(ORDER BY reportDate)
        ) * 100 AS percent_change_min_price
    FROM charli_xcx_events
    WHERE vs_event_id = '3768211'
)
SELECT
    name,
    venue,
    Days_Until_Event,
    ticketCount,
    ROUND(percent_change_ticket_count, 2) AS percent_change_ticket_count,
    minPrice,
    ROUND(percent_change_min_price, 2) AS percent_change_min_price
FROM charli_xcx_boston;

 * mysql://admin:***@tickets.ceaw7c57s0vu.us-east-1.rds.amazonaws.com/tickets
9 rows affected.


[('Charli XCX', 'House of Blues Boston', 12, 116, None, 115, None),
 ('Charli XCX', 'House of Blues Boston', 9, 104, Decimal('-10.34'), 125, Decimal('8.70')),
 ('Charli XCX', 'House of Blues Boston', 8, 98, Decimal('-5.77'), 138, Decimal('10.40')),
 ('Charli XCX', 'House of Blues Boston', 8, 102, Decimal('4.08'), 138, Decimal('0.00')),
 ('Charli XCX', 'House of Blues Boston', 6, 100, Decimal('-1.96'), 123, Decimal('-10.87')),
 ('Charli XCX', 'House of Blues Boston', 4, 82, Decimal('-18.00'), 96, Decimal('-21.95')),
 ('Charli XCX', 'House of Blues Boston', 3, 67, Decimal('-18.29'), 103, Decimal('7.29')),
 ('Charli XCX', 'House of Blues Boston', 2, 59, Decimal('-11.94'), 98, Decimal('-4.85')),
 ('Charli XCX', 'House of Blues Boston', 0, 34, Decimal('-42.37'), 73, Decimal('-25.51'))]

### From about two weeks out to one week out, we see a decrease in ticket count and an increase in the minimum price. With 8 days until the event, it reached as high as 138. However, from this point on until the event we see an even greater decrease in ticket count but we also observe the minimum price decreasing as well. Based on these results, it seems like selling a week out from the day of the event would have netted you the most profit. This information can be used in the future if Charli XCX tours again and plays in a market similar to Boston.

# Secondary Questions
## Was this event profitable?
### Another important question that you have to ask yourself is whether or not you should buy an event. Instead of blindly making a 50/50 call, you can rely on historical data to make a better informed decision. This query compares the average minimum resale price from Vividseats to the minimum face value price from Ticketmaster to determine whether an event was profitable.
#### SQL features used: GROUP BY, JOIN, CASE

In [16]:
%%sql
SELECT
    vi.name,
    t.genre,
    vi.venue,
    ROUND(AVG(vi.minPrice)) AS average_resale_get_in,
    t.minPrice AS retail_get_in,
    CASE WHEN AVG(vi.minPrice) > t.minPrice THEN 1 ELSE 0 END AS profitable
FROM vividseats vi
JOIN venue ve
    ON vi.venue = ve.name
JOIN ticketmaster t
    ON ve.name = t.venue
GROUP BY vs_event_id
ORDER BY vi.eventDate, vi.name, reportDate;

 * mysql://admin:***@tickets.ceaw7c57s0vu.us-east-1.rds.amazonaws.com/tickets
35 rows affected.


[('Bon Jovi', 'Country', 'Enmarket Arena', Decimal('52'), 56, 0),
 ('Jawbreaker', 'Pop', 'House of Blues Chicago', Decimal('11'), 28, 0),
 ('Bon Jovi', 'Rock', 'Moody Center ATX', Decimal('106'), 60, 1),
 ('Netflix Is A Joke Fest - Bill Burr', 'Comedy', 'Hollywood Palladium', Decimal('66'), 64, 1),
 ('Clinton Kane', 'Pop', 'The Bronze Peacock at House of Blues Houston', Decimal('45'), 25, 1),
 ('Paul McCartney', 'Rock', 'Oakland Arena', Decimal('106'), 30, 1),
 ('Pearl Jam (Rescheduled from 4/15/2020)', 'Comedy', 'Kia Forum', Decimal('116'), 60, 1),
 ('Garth Brooks', 'Country', 'Notre Dame Stadium', Decimal('78'), 92, 0),
 ('iHeartCountry Festival - Thomas Rhett, Carrie Underwood, Zac Brown Band', 'Rock', 'Moody Center ATX', Decimal('109'), 60, 1),
 ('Mxmtoon', 'Pop', 'House of Blues Chicago', Decimal('8'), 28, 0),
 ('Pearl Jam (Rescheduled from 4/16/2020)', 'Comedy', 'Kia Forum', Decimal('129'), 60, 1),
 ('Dave Matthews Band', 'Rock', 'Moody Center ATX', Decimal('67'), 60, 1),
 ('Dave Matthews Band', 'Rock', 'Dos Equis Pavilion', Decimal('74'), 50, 1),
 ('Garth Brooks', 'Country', 'Paul Brown Stadium', Decimal('77'), 85, 0),
 ('Twice', 'Pop', 'Banc of California Stadium', Decimal('119'), 60, 1),
 ('Dave Matthews Band', 'Rock', 'Brandon Amphitheater', Decimal('91'), 50, 1),
 ('Paul McCartney', 'Rock', 'Dickies Arena', Decimal('404'), 40, 1),
 ('Rex Orange County', 'Rock', 'Germania Insurance Amphitheater', Decimal('55'), 30, 1),
 ('Garth Brooks', 'Country', 'Gillette Stadium', Decimal('47'), 85, 0),
 ('Kane Brown', 'Country', 'Amalie Arena', Decimal('62'), 47, 1),
 ('Luke Combs', 'Country', 'Empower Field at Mile High', Decimal('76'), 21, 1),
 ('Nine Inch Nails', 'Rock', 'The Met Philadelphia', Decimal('86'), 50, 1),
 ('Kane Brown', 'Country', 'Enmarket Arena', Decimal('49'), 56, 0),
 ('Keshi', 'Other', 'Theatre of Living Arts', Decimal('102'), 30, 1),
 ('Coldplay', 'Rock', 'Soldier Field', Decimal('77'), 34, 1),
 ('Paul McCartney', 'Rock', 'Thompson-Boling Arena', Decimal('138'), 70, 1),
 ('Keshi', 'Pop', 'House of Blues Chicago', Decimal('90'), 28, 1),
 ('Morgan Wallen', 'Country', 'PNC Music Pavilion', Decimal('282'), 60, 1),
 ('Coldplay', 'Rock', 'MetLife Stadium', Decimal('72'), 40, 1),
 ('Chris Rock', 'Comedy', 'Taft Theatre', Decimal('62'), 50, 1),
 ('Jimmy Buffett', 'Rock', 'Moody Center ATX', Decimal('18'), 60, 0),
 ('Paul McCartney', 'Rock', 'MetLife Stadium', Decimal('107'), 40, 1),
 ('Rex Orange County', 'Rock', 'BMO Harris Pavilion', Decimal('44'), 30, 1),
 ('Keshi', 'Other', 'The Observatory North Park', Decimal('94'), 30, 1),
 ('Rex Orange County', 'Rock', 'Huntington Bank Pavilion at Northerly Island', Decimal('50'), 30, 1)]

### These results can be used in the future to make better informed buying decisions. For instance, the average resale get in price (106) for Paul McCartney's show at Oakland Arena is a lot higher than the retail get in price (30). If he were to ever play a show again at that venue or in that area (SF Bay Area) I would feel confident buying that show based on how his Oakland Arena show is doing.

## Can we predict profitability based on genre?
### This query builds off of the previous one. To possibly help make our decision-making process easier, we can calculate the percentage of profitable events based on genre.
#### SQL features used: GROUP BY, JOIN, CASE, CTE

In [21]:
%%sql
WITH profitable AS (
    SELECT
        vi.name,
        t.genre,
        vi.venue,
        ROUND(AVG(vi.minPrice)) AS average_resale_get_in,
        t.minPrice AS retail_get_in,
        CASE WHEN AVG(vi.minPrice) > t.minPrice THEN 1 ELSE 0 END AS profitable
    FROM vividseats vi
    JOIN venue ve
        ON vi.venue = ve.name
    JOIN ticketmaster t
        ON ve.name = t.venue
    GROUP BY vs_event_id
    ORDER BY vi.eventDate, vi.name, reportDate
),
profitable_percentage AS (
    SELECT
        genre,
        (((COUNT(CASE WHEN profitable = 1 AND genre = 'Other' THEN profitable ELSE NULL END)) / COUNT(genre)) * 100) AS profitable_other,
        ((COUNT(CASE WHEN profitable = 1 AND genre = 'Comedy' THEN profitable ELSE NULL END)) / COUNT(genre)) * 100 AS profitable_comedy,
        ((COUNT(CASE WHEN profitable = 1 AND genre = 'Pop' THEN profitable ELSE NULL END)) / COUNT(genre)) * 100 AS profitable_pop,
        ((COUNT(CASE WHEN profitable = 1 AND genre = 'Country' THEN profitable ELSE NULL END)) / COUNT(genre)) * 100 AS profitable_country,
        ((COUNT(CASE WHEN profitable = 1 AND genre = 'Rock' THEN profitable ELSE NULL END)) / COUNT(genre)) * 100 AS profitable_rock
    FROM profitable
    GROUP BY genre
)
SELECT
    genre,
    FORMAT(profitable_other, 2) AS other_profitability,
    FORMAT(profitable_comedy, 2) AS comdey_profitability,
    FORMAT(profitable_pop, 2) AS pop_profitability,
    FORMAT(profitable_country, 2) AS country_profitability,
    FORMAT(profitable_rock, 2) AS rock_profitability
FROM profitable_percentage;

 * mysql://admin:***@tickets.ceaw7c57s0vu.us-east-1.rds.amazonaws.com/tickets
5 rows affected.


[('Country', '0.00', '0.00', '0.00', '37.50', '0.00'),
 ('Pop', '0.00', '0.00', '60.00', '0.00', '0.00'),
 ('Rock', '0.00', '0.00', '0.00', '0.00', '93.75'),
 ('Comedy', '0.00', '100.00', '0.00', '0.00', '0.00'),
 ('Other', '100.00', '0.00', '0.00', '0.00', '0.00')]

### The results of this query tells us that country and pop events are not as profitable as rock and comedy. Of course, we are working with a very limited sample of data, but I can see how this query can prove to be more useful and reliable in the future if more events are added to the database.